mongo-connector stops working when active mongos and node go offline

Our environment has following MongoDB cluster setup:
**Site 1:**
config1: config & mongos
config2: config & mongos
node1: rs1 & rs2
node2: rs1 & rs2

**Site 3:**
config3: config & mongos
node3: rs1 & rs2

mongo-connector **mainAddress** is "config1:27017,config2:27017,config3:27017"

On the weekend **Site 1** went down and node3 on **Site 2** was elected primary. During this time mongo-connector raised an AutoReconnect and stopped indexing documents. The mongo-connector process is still alive, but nothing is getting inserted to Solr.

Judging from where the exception was raised from, it would seem that the OplogThread noticed the disconnection and the control returned to Connector. However, the mongos Connector was connected to (this I am guessing but it would make sense) is also down, because it resides on the same site. In this scenario, the main connection should probably be reconnected?

[mongo-connector.txt](https://github.com/mongodb-labs/mongo-connector/files/521814/mongo-connector.txt)

Edit: Including the traceback:
```python
2016-10-09 00:53:06,512 [ERROR] mongo_connector.util:92 - Fatal Exception
Traceback (most recent call last):
  File "/usr/lib/python2.6/site-packages/mongo_connector-2.4.1-py2.6.egg/mongo_connector/util.py", line 90, in wrapped
    func(*args, **kwargs)
  File "/usr/lib/python2.6/site-packages/mongo_connector-2.4.1-py2.6.egg/mongo_connector/connector.py", line 316, in run
    for shard_doc in main_conn['config']['shards'].find():
  File "build/bdist.linux-x86_64/egg/pymongo/cursor.py", line 1149, in next
    if len(self.__data) or self._refresh():
  File "build/bdist.linux-x86_64/egg/pymongo/cursor.py", line 1081, in _refresh
    self.__codec_options.uuid_representation))
  File "build/bdist.linux-x86_64/egg/pymongo/cursor.py", line 996, in __send_message
    res = client._send_message_with_response(message, **kwargs)
  File "build/bdist.linux-x86_64/egg/pymongo/mongo_client.py", line 1366, in _send_message_with_response
    raise AutoReconnect(str(e))
AutoReconnect: [Errno 113] No route to host
```

--- a/mongo_connector/connector.py
+++ b/mongo_connector/connector.py
@@ -31,7 +31,7 @@ from mongo_connector.oplog_manager import OplogThread
 from mongo_connector.doc_managers import doc_manager_simulator as simulator
 from mongo_connector.doc_managers.doc_manager_base import DocManagerBase
 from mongo_connector.command_helper import CommandHelper
-from mongo_connector.util import log_fatal_exceptions
+from mongo_connector.util import log_fatal_exceptions, retry_until_ok
 from mongo_connector.dest_mapping import DestMapping
 
 from pymongo import MongoClient
@@ -330,7 +330,8 @@ class Connector(threading.Thread):
         else:       # sharded cluster
             while self.can_run is True:
 
-                for shard_doc in main_conn['config']['shards'].find():
+                for shard_doc in retry_until_ok(
+                        lambda: list(main_conn.config.shards.find())):
                     shard_id = shard_doc['_id']
                     if shard_id in self.shard_set:
                         shard_thread = self.shard_set[shard_id]

--- a/mongo_connector/connector.py
+++ b/mongo_connector/connector.py
@@ -336,8 +336,8 @@ class Connector(threading.Thread):
         else:       # sharded cluster
             while self.can_run is True:
 
-                for shard_doc in retry_until_ok(
-                        lambda: list(self.main_conn.config.shards.find())):
+                for shard_doc in retry_until_ok(self.main_conn.admin.command,
+                                                'listShards')['shards']:
                     shard_id = shard_doc['_id']
                     if shard_id in self.shard_set:
                         shard_thread = self.shard_set[shard_id]

--- a/mongo_connector/connector.py
+++ b/mongo_connector/connector.py
@@ -59,6 +59,9 @@ class Connector(threading.Thread):
         # main address - either mongos for sharded setups or a primary otherwise
         self.address = mongo_address
 
+        # connection to the main address
+        self.main_conn = None
+
         # List of DocManager instances
         if doc_managers:
             self.doc_managers = doc_managers
@@ -270,25 +273,31 @@ class Connector(threading.Thread):
                     (name, util.long_to_bson_ts(timestamp))
                     for name, timestamp in data)
 
+    def create_authed_client(self, address=None, **kwargs):
+        kwargs.update(self.ssl_kwargs)
+        if address is None:
+            address = self.address
+        client = MongoClient(address, tz_aware=self.tz_aware, **kwargs)
+        if self.auth_key is not None:
+            client['admin'].authenticate(self.auth_username, self.auth_key)
+        return client
+
     @log_fatal_exceptions
     def run(self):
         """Discovers the mongo cluster and creates a thread for each primary.
         """
-        main_conn = MongoClient(
-            self.address, tz_aware=self.tz_aware, **self.ssl_kwargs)
-        if self.auth_key is not None:
-            main_conn['admin'].authenticate(self.auth_username, self.auth_key)
+        self.main_conn = self.create_authed_client()
         self.read_oplog_progress()
         conn_type = None
 
         try:
-            main_conn.admin.command("isdbgrid")
+            self.main_conn.admin.command("isdbgrid")
         except pymongo.errors.OperationFailure:
             conn_type = "REPLSET"
 
         if conn_type == "REPLSET":
             # Make sure we are connected to a replica set
-            is_master = main_conn.admin.command("isMaster")
+            is_master = self.main_conn.admin.command("isMaster")
             if "setName" not in is_master:
                 LOG.error(
                     'No replica set at "%s"! A replica set is required '
@@ -297,20 +306,17 @@ class Connector(threading.Thread):
                 return
 
             # Establish a connection to the replica set as a whole
-            main_conn.close()
-            main_conn = MongoClient(
-                self.address, replicaSet=is_master['setName'],
-                tz_aware=self.tz_aware, **self.ssl_kwargs)
-            if self.auth_key is not None:
-                main_conn.admin.authenticate(self.auth_username, self.auth_key)
+            self.main_conn.close()
+            self.main_conn = self.create_authed_client(
+                replicaSet=is_master['setName'])
 
             # non sharded configuration
             oplog = OplogThread(
-                main_conn, self.doc_managers, self.oplog_progress,
+                self.main_conn, self.doc_managers, self.oplog_progress,
                 self.dest_mapping, **self.kwargs)
             self.shard_set[0] = oplog
             LOG.info('MongoConnector: Starting connection thread %s' %
-                     main_conn)
+                     self.main_conn)
             oplog.start()
 
             while self.can_run:
@@ -331,7 +337,7 @@ class Connector(threading.Thread):
             while self.can_run is True:
 
                 for shard_doc in retry_until_ok(
-                        lambda: list(main_conn.config.shards.find())):
+                        lambda: list(self.main_conn.config.shards.find())):
                     shard_id = shard_doc['_id']
                     if shard_id in self.shard_set:
                         shard_thread = self.shard_set[shard_id]
@@ -358,11 +364,8 @@ class Connector(threading.Thread):
                             dm.stop()
                         return
 
-                    shard_conn = MongoClient(
-                        hosts, replicaSet=repl_set, tz_aware=self.tz_aware,
-                        **self.ssl_kwargs)
-                    if self.auth_key is not None:
-                        shard_conn['admin'].authenticate(self.auth_username, self.auth_key)
+                    shard_conn = self.create_authed_client(
+                        hosts, replicaSet=repl_set)
                     oplog = OplogThread(
                         shard_conn, self.doc_managers, self.oplog_progress,
                         self.dest_mapping, **self.kwargs)

--- a/tests/test_connector_sharded.py
+++ b/tests/test_connector_sharded.py
@@ -14,6 +14,7 @@
 
 import os
 import sys
+import time
 
 sys.path[0:0] = [""]
 
@@ -29,34 +30,63 @@ from tests import unittest, SkipTest
 class TestConnectorSharded(unittest.TestCase):
 
     def setUp(self):
-        if not (db_user and db_password):
-            raise SkipTest('Need to set a user/password to test this.')
+        if db_user and db_password:
+            auth_args = dict(auth_username=db_user, auth_key=db_password)
+        else:
+            auth_args = {}
         self.cluster = ShardedClusterSingle().start()
+        self.dm = DocManager()
+        self.connector = Connector(
+            mongo_address=self.cluster.uri,
+            doc_managers=[self.dm],
+            **auth_args
+        )
+        self.connector.start()
 
     def tearDown(self):
+        self.connector.join()
         try:
             os.unlink('oplog.timestamp')
         except OSError:
             pass
         self.cluster.stop()
 
-    def test_start_with_auth(self):
-        dm = DocManager()
-        connector = Connector(
-            mongo_address=self.cluster.uri,
-            doc_managers=[dm],
-            auth_username=db_user,
-            auth_key=db_password
-        )
-        connector.start()
 
+class TestConnectorShardedMongosFailure(TestConnectorSharded):
+
+    def test_mongos_connection_failure(self):
+        """Test that the connector handles temporary mongos failure"""
+        client = self.cluster.client()
+        coll = client.test.test
+        coll.insert_one({'doc': 1})
+        assert_soon(lambda: len(self.dm._search()) == 1)
+
+        # Temporarily shutdown the connector mongos connection
+        self.connector.main_conn.close()
+        time.sleep(5)
+        coll.insert_one({'doc': 2})
+        assert_soon(lambda: len(self.dm._search()) == 2)
+
+        # Bring mongos back online
+        self.connector.main_conn = self.connector.create_authed_client()
+        coll.insert_one({'doc': 3})
+        assert_soon(lambda: len(self.dm._search()) == 3)
+        client.close()
+
+
+class TestConnectorShardedAuth(TestConnectorSharded):
+
+    def setUp(self):
+        if not (db_user and db_password):
+            raise SkipTest('Need to set a user/password to test this.')
+        super(TestConnectorShardedAuth, self).setUp()
+
+    def test_start_with_auth(self):
         # Insert some documents into the sharded cluster.  These
         # should go to the DocManager, and the connector should not
         # have an auth failure.
         self.cluster.client().test.test.insert_one({'auth_failure': False})
-        assert_soon(lambda: len(dm._search()) > 0)
-
-        connector.join()
+        assert_soon(lambda: len(self.dm._search()) > 0)
 
 
 if __name__ == '__main__':

